### ISSUE TITLE
Restoring backup to Xenserver pool is not working as expected.

### ISSUE THREAD

**Issue by mkaliyannan** [izderadicka/xapi-back#6]
My  current setup is 2 xenserver .

Version 6.5 patches with SP1 upto date.
Storage : 2 local storage ( named as local1 for server 1 followed by local 2 for server2)  + 1 NFS storage repository ( defaul storage)

Scenario : 

Server 1 is have running a  VM name called  CentOS-VM. ( VM disk is stored on local1 storage)
Server 2 is have running a VM name called Ubuntu-VM.     ( VM disk is stored on local2 storage)

If  I run a backup through your tool, it backup successfully either manual or automatic.

Restore : 

(1)  for instance If  I restore a CentOS-VM using  xb restore, it was restored to NFS storage and shown as template . Actually, It suppose to be restore on local1 storage and should be restore as VM. Not as a template. ( check the picture below)

(2)  If I removed the NFS storage from the pool then If I  took a backup and restore using your tool it works as expected !! ( restore properly to corresponding storage) 

Problem : Only the problem was happen if  I use the shared repository on the pool. This needs to be fix.,

The reason need to be fix is If the backups are stored in nfs, it will take long time to copy to original disk from the template  so this must be avoided. 

![screenshot_1](https://cloud.githubusercontent.com/assets/8954012/12214993/4712bbb4-b666-11e5-9055-63ab470605b3.png)

The above mentioned arrow mark is pointing that backup is storing as a template.

**Comment by mkaliyannan** [comment]
I never restored the backup from remote machine using xe vm-import tool. 

Can you put the syntax here ?

I normally copy the backup from remote machine to local machine through ssh then I restore using xe vm-import filename= sr-uuid=.

**Comment by izderadicka** [comment]
Hi Mohan,
first thanks for all the hard work with testing up to now.
I think it goes down to the fact that restore is not specifying SR and uses default SR - which in your case was shared NFS storage.    
Obviously there are some additional actions on the XenServer side -   because then VM appears as a template.  This would be probably result of SR change.
I think the right solution will be to add possibility to specify SR during restore, what do you think?

As for xe -   it really does not matter from where you are running it ( but you can install it basically on any linux machine) -   if using sr-uuid  does provide expected result, that's the way to go.

Also you can try restore as described here:  http://xapi-project.github.io/xen-api/importexport.html  with curl -T  and appropriate URL params.   Because this is exactly what is happening on background in xb - image is sent via PUT HTTP method to server.

**Comment by mkaliyannan** [comment]
I think you are right. Specifying SR would probably solve the problem.

Sent from my iPhone

> On Jan 9, 2016, at 1:21 AM, Ivan Zderadicka <notifications@github.com> wrote:
> 
> Hi Mohan,
> first thanks for all the hard work with testing up to now.
> I think it goes down to the fact that restore is not specifying SR and uses default SR - which in your case was shared NFS storage.
> 
> Obviously there are some additional actions on the XenServer side - because then VM appears as a template. This would be probably result of SR change.
> I think the right solution will be to add possibility to specify SR during restore, what do you think?
> 
> As for xe - it really does not matter from where you are running it ( but you can install it basically on any linux machine) - if using sr-uuid does provide expected result, that's the way to go.
> 
> Also you can try restore as described here: http://xapi-project.github.io/xen-api/importexport.html with curl -T and appropriate URL params. Because this is exactly what is happening on background in xb - image is sent via PUT HTTP method to server.
> 
> —
> Reply to this email directly or view it on GitHub.
> 

Disclaimer in plain tex

**Comment by mkaliyannan** [comment]
Cool. I tried as your advice. I followed as described here: http://xapi-project.github.io/xen-api/importexport.html . 

I stopped the VM tried to export and import it works fine. Still it looks Default storage.

curl -T Netscaler http://root:xxxxx@10.10.50.202/import?sr_id=b011b7ac-787e-cf24-247b-3bfba62293d5

I tried to define storage repo as above, but gave me 500 error as below . If I not defined sr_id it goes back to restore with default storage. 

**********************************************
HTTP 500 internal server error

An unexpected error occurred; please wait a while and try again. If the problem persists, please contact your support representative.
Additional information

NO_HOSTS_AVAILABLE: [ ][

**Comment by izderadicka** [comment]
I'd recommend to look at log file on host - there might be more info why import failed.

And actually xb already has parameter for sr_id - so you might try it as well.  

```
 xb restore -h
usage: xb restore [-h] --host HOST --vm VM [--no-progress] [--restore]
                  [--sr_id SR_ID]

optional arguments:
  -h, --help     show this help message and exit
  --host HOST    name of host to connect to (from config file)
  --vm VM        Name of VM
  --no-progress  Do not print progress
  --restore      Restore as replacement of original VM (MAC is same)
  --sr_id SR_ID  uuid of SR to import to (if other then default SR)
```

**Comment by mkaliyannan** [comment]
Sorry for being late reply.  I already tried the sr_id parameter but it doesn't work...

**Comment by mkaliyannan** [comment]
2016-01-11 15:49:32,514 ERROR:Command restore failed with [Errno 104] Connection reset by peer

Tha's the above log when i use sr_id parameter

**Comment by izderadicka** [comment]
This is the xb log , right?

What is in Xenserver log  on master - especially /var/log/xensource.log 
would be most useful - there should be all XAPI related messages

Where is the storage you trying to use?
Is it shared?
Is it local?  Is it local on master or on other host in the pool?

I think problem might be that xb is now restoring to master only -   if 
storage is on the other host we might need to post to other host?  I'll 
look into it later - for next few days I might be busy on other things.

I.

**Comment by izderadicka** [comment]
I've pushed fix to master branch - it should be able now to restore to given SR uuid.  Also there is new command `xb srs`, which should show available SRs on a given host.

Concerning a VM to be restored as a template that's a different story.  If backup is done from running VM ( as a snapshot) XAPI makes this snapshot  a template.  (You can make always backup with  --shutdown,  which will do cold backup of VM that is also restored as VM) If you need other behaviour please open another issues  for it.

Please let me know if it works for you.

**Comment by izderadicka** [comment]
I believe issue is fixed now with e17780e5b7a13a0c28576e76bed17a879cd8a3df.
Also new switch added to restore command -   --as-vm - thus snapshots can be restored as VM directly - dbf799b755d7d0013775b38108bdaa7b0b68476a
